### PR TITLE
Skip batches fast with accelerate

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR uses the lastest from `Accelerate` to quickly skip batches when resuming training (it's only going to be quicker for a regular dataset, iterable datasets will still require a manual pass though the first batches).

Note that the RNG seeds can't be reloaded before we have started the iteration of the data loader because the random sampler used by the data laoder needs to use the seed at this stage, so there is a small path to load it after the iteration has started

cc @stas00 if you want to experiment (needs to use Accelerate main until v0.16 is out).

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -193,6 +194,15 @@
...
 
 
+skip_first_batches = None
+if is_accelerate_available():
+    from accelerate import __version__ as accelerate_version
+
+    # Change this to >= 0.16 once the release is out
+    if version.parse(accelerate_version) > version.parse("0.15"):
+        from accelerate import skip_first_batches
+
```
Comment (stas00) [inline src/transformers/trainer.py:204]:
```suggestion
    if version.parse(accelerate_version) > version.parse("0.15"):
        from accelerate import skip_first_batches
    else:
        skip_first_batches = None
```

no?

Comment (sgugger) [reply src/transformers/trainer.py:204]: It's set before the first test of accelerate availability as it also needs to be None if accelerate is not available.

Comment (stas00) [reply src/transformers/trainer.py:204]: ah, oops, I missed it! sorry about that.

Comment (sgugger) [reply src/transformers/trainer.py:204]: No worries!

**Comment by stas00** [comment]
Thank you for implementing this, Sylvain,

Is this PR going to sit for a bit until the new Accelerate comes out? 

Currently I have to work on some other urgent things and SLURM makes it hard to quickly test things, but would be happy to test once I get the opportunity.

**Comment by sgugger** [comment]
The new release is just out actually, so this should be merged as soon as positively reviewed :-)

**Comment by stas00** [comment]
ok, great! congrats!

let me then just do a quick test on my desktop.

**Review by stas00 — APPROVED** [review]
OK, I did only a small test that exercised the new code path - seems to work fine.

I will be able to do a est on a large dataset probably some time in the next few days.

**Review by LysandreJik — APPROVED** [review]
Thank you @sgugger

**Comment by stas00** [comment]
Hmm, on a small test it worked fine, but I see it fails on the real training.

snapshot_88

The straight line is before this change, the restarting one is after.

I was training with `--log_level warning` so didn't get the info log.

Any thoughts to what might have gone wrong?

The example script that was running is here:  https://github.com/huggingface/m4/pull/922/files#diff-27452e2e8d112cbfd59f7900ef3b39dc35a4a0faf2d967fd86e399fe7ccb1ba2R192-R217

I have the latest released `accelerate` and `transformers@main` (from Feb 1)

**Comment by stas00** [comment]
On the other hand it finished training at the same time as before, so the dataloader issues `StopIteration` as before.

which means some accounting was off - that is epoch isn't calculated correctly.

**Comment by sgugger** [comment]
The training should finish at the same loss (this is tested with and without randomness in the model for small training), so normally you're good with the data. I think the problem lies with [this line](https://github.com/huggingface/transformers/blob/182afb7dc6f40aea5f5bb41710cb5207d187b022/src/transformers/trainer.py#L1909) which uses the `step` variable which now goes from 0 to len(data_loader) - num_batches_skipped (instead of num_batch_skipped to len(data_loader) before my PR). Will push a fix today or tomorrow!

**Comment by stas00** [comment]
super! thank you, Sylvain.

**Comment by sgugger** [comment]
Should be fixed by the PR mentioned above.

**Pull request merged**